conjure-http/src/server.rs

```diff
@@ -300,3 +327,334 @@ impl<'a> RequestContext<'a> {
+        Self::check_content_type(headers)?;
+        let buf = private::async_read_body(body, Some(SERIALIZABLE_REQUEST_SIZE_LIMIT)).await?;
+        json::server_from_slice(&buf).map_err(|e| Error::service(e, InvalidArgument::new()))
+    }
+}
+
+/// A trait implemented by response serializers used by custom Conjure server trait implementations.
+pub trait SerializeResponse<T, W> {
+    /// Serializes the response.
+    fn serialize(request_headers: &HeaderMap, value: T)
+        -> Result<Response<ResponseBody<W>>, Error>;
+}
+
+/// A trait implemented by response serializers used by custom async Conjure server trait
+/// implementations.
+pub trait AsyncSerializeResponse<T, W> {
+    /// Serializes the response.
+    fn serialize(
+        request_headers: &HeaderMap,
+        value: T,
+    ) -> Result<Response<AsyncResponseBody<W>>, Error>;
+}
+
+/// A serializer which encodes `()` as an empty body and status code of `204 No Content`.
+pub enum EmptyResponseSerializer {}
+
+impl EmptyResponseSerializer {
+    fn serialize_inner<T>(body: T) -> Result<Response<T>, Error> {
+        let mut response = Response::new(body);
+        *response.status_mut() = StatusCode::NO_CONTENT;
+        Ok(response)
+    }
+}
+
+impl<W> SerializeResponse<(), W> for EmptyResponseSerializer {
+    fn serialize(_: &HeaderMap, _: ()) -> Result<Response<ResponseBody<W>>, Error> {
+        Self::serialize_inner(ResponseBody::Empty)
+    }
+}
+
+impl<W> AsyncSerializeResponse<(), W> for EmptyResponseSerializer {
+    fn serialize(_: &HeaderMap, _: ()) -> Result<Response<AsyncResponseBody<W>>, Error> {
+        Self::serialize_inner(AsyncResponseBody::Empty)
+    }
+}
+
+/// A serializer which acts like a Conjure-generated client would.
+pub enum ConjureResponseSerializer {}
+
+impl ConjureResponseSerializer {
+    fn serialize_inner<T, B>(
+        value: T,
+        make_body: impl FnOnce(Bytes) -> B,
+    ) -> Result<Response<B>, Error>
+    where
+        T: Serialize,
+    {
+        let body = json::to_vec(&value).map_err(Error::internal)?;
+
+        let mut response = Response::new(make_body(body.into()));
+        response
+            .headers_mut()
+            .insert(CONTENT_TYPE, APPLICATION_JSON);
+
+        Ok(response)
+    }
+}
+
+impl<T, W> SerializeResponse<T, W> for ConjureResponseSerializer
+where
+    T: Serialize,
+{
+    fn serialize(
+        _request_headers: &HeaderMap,
+        value: T,
+    ) -> Result<Response<ResponseBody<W>>, Error> {
+        Self::serialize_inner(value, ResponseBody::Fixed)
+    }
+}
+
+impl<T, W> AsyncSerializeResponse<T, W> for ConjureResponseSerializer
+where
+    T: Serialize,
+{
+    fn serialize(
+        _request_headers: &HeaderMap,
+        value: T,
+    ) -> Result<Response<AsyncResponseBody<W>>, Error> {
+        Self::serialize_inner(value, AsyncResponseBody::Fixed)
+    }
+}
+
+/// A trait implemented by header decoders used by custom Conjure server trait implementations.
+pub trait DecodeHeader<T> {
```

Comment:
For my own understanding, when should a sort of "static method" approach like this be preferred over something like,

```
pub trait FromHeader {
  fn decode(headers ...) -> Self
}
```
where the deserialization of a type is sort of "baked into" that type?

Reply:
That would mean there can only be one way to parse a header into some type, and mean that the only place any implementations for stdlib types could be defined is in conjure-http.

Reply:
For example, you may have an API that needs to hex-encode an integer rather than decimal-encoding it. Using separate decoder types allows the method itself to take a raw `i32` in both cases and just have differing annotations.

Reply:
Ah, that makes sense! I was thinking that users would create newtypes in that case, but that would get pretty annoying

Reply:
Yeah I figure it's better to add annotation burden rather than extra code at every call site.